### PR TITLE
RFC: Add support for determining name of script

### LINKED ISSUE #14109: Name of .jl Script (aka. $0)
When running Julia programs from the command line `julia program.jl` I would like to be able to determine the name of the file Julia is executing. Note that this is different from `@__FILE__` as I may want to find out the script name in an included file.

In Bash and [Perl](http://perlmaven.com/argv-in-perl) the name of the script can be referenced through $0. I think in Julia it could be reference via `FILE_NAME` or `SCRIPT_NAME`.

### PR DESCRIPTION
**Pull request by omus**
Fixes  #14109.

### CODE REVIEW

File: base/exports.jl [hunk 1]
```diff
@@ -174,6 +174,7 @@ export
     UnicodeError,
 
 # Global constants and variables
+    FILE_NAME,
```
Comment (tkelman) [inline base/exports.jl]: this should be alphabetized in each grouping

Comment (omus) [inline base/exports.jl]: Good to know. Thanks

**Comment by sjkelly** [comment]
I think n+1 people have wanted this, so docs and a NEWS.md entry might be in order?

(I also wonder if this opens up module as a script functionality?)

**Comment by omus** [comment]
I added/updated documentation and an entry to NEWS.md. 

What do people think of the name `FILE_NAME` to indicate the name of the script? Anyone have any better suggestions? Also, `FILE_NAME` is currently a non-constant global. Is this a major problem or is there a good way to solve this that I'm not thinking of?

**Comment by KristofferC** [comment]
Maybe use `const FILE_NAME = Ref{UTF8String}()`?

**Comment by iamed2** [comment]
@omus:
CLI names for the script file:

Julia: `programfile`
Ruby: `programfile`
Python: `file`
Lua: `script`
Guile: `file`
Racket: `file`
Perl: `programfile`
Bash: `script-file`
Fish: `help`
R: `file`
PHP: `file`

Seems like `PROGRAMFILE` might be the best, in order to be consistent with the CLI help.

**Comment by omus** [comment]
I like that name better. I'll change it to `PROGRAM_FILE` to make it slightly more readable.

**Comment by omus** [comment]
After messing around with this more `ARGS` only has the program file removed right before the program file is run. That means that for code executing from using `--load`, `--eval`, or `--print` will not be able to determine the `PROGRAM_FILE` since it will still be included within `ARGS`. I was planning to initialize `PROGRAM_FILE` as a constant by initializing it at the last possible moment via:
```julia
global const PROGRAM_FILE = UTF8String(shift!(ARGS))
```
Unfortunately this won't work as I would also like to support having `PROGRAM_FILE` defined and usable to any external script passed into Julia.

The repercussions of this is that `PROGRAM_FILE` cannot be a constant and will only be assigned correctly to a filename from within the `PROGRAM_FILE` itself and included files.

File: NEWS.md [hunk 2]
```diff
@@ -151,6 +151,8 @@ New language features
 
   * `@__LINE__` special macro now available to reflect invocation source line number ([#12727]).
 
+  * `PROGRAM_FILE` global is now available for determining the name of the running script ([#14114]).
```
Comment (stevengj) [inline NEWS.md]: you need to run `julia doc/NEWS-update.jl` to update the cross-reference links.

File: doc/stdlib/constants.rst [hunk 3]
```diff
@@ -14,6 +14,10 @@ Constants
    A symbol representing the name of the operating system. Possible values
    are ``:Linux``, ``:Darwin`` (OS X), or ``:Windows``.
 
+.. data:: FILE_NAME
```
Comment (stevengj) [inline doc/stdlib/constants.rst]: PROGRAM_FILE, not FILE_NAME?

Comment (omus) [inline doc/stdlib/constants.rst]: I missed this one during the rename. Thanks

File: NEWS.md (hunk 2 above)
Comment (omus) [inline NEWS.md]: Ok. Is there some Julia contribution documentation that mentions this?

Comment (tkelman) [inline NEWS.md]: if it isn't mentioned somewhere in `CONTRIBUTING.md` it probably should be

**Comment by omus** [comment]
I'll squash these commits

**Comment by tkelman** [comment]
was the issue with
> The repercussions of this is that `PROGRAM_FILE` cannot be a constant and will only be assigned correctly to a filename from within the `PROGRAM_FILE` itself and included files.

sorted, or is it a minor concern?

**Comment by omus** [comment]
@tkelman I think it is a minor concern. Basically `PROGRAM_FILE` will be empty for the `~/.juliarc` and the `--load` file. I could sort this problem out but that the change would involve taking the PROGRAM_FILE out of ARGS prior to running either of these file. Changing that behaviour could break things:

Here's a demonstration:

```bash
$ echo 'println("PROGRAM_FILE: $PROGRAM_FILE"); for (i, arg) in enumerate(ARGS); println("$i: $arg"); end' > script.jl

$ ./julia script.jl foo -bar --baz
PROGRAM_FILE: script.jl
1: foo
2: -bar
3: --baz

$ ./julia -L script.jl script.jl foo -bar --baz
PROGRAM_FILE: 
1: script.jl
2: foo
3: -bar
4: --baz
PROGRAM_FILE: script.jl
1: foo
2: -bar
3: --baz
```
What do you think?

**Comment by tkelman** [comment]
I guess it not being `const` means people can modify it which could be confusing or abused?

File: doc/stdlib/constants.rst [hunk 4]
```diff
@@ -14,6 +14,10 @@ Constants
    A symbol representing the name of the operating system. Possible values
    are ``:Linux``, ``:Darwin`` (OS X), or ``:Windows``.
 
+.. data:: PROGRAM_FILE
+
+   A string containing the script name passed to Julia.
```
Comment (tkelman) [inline doc/stdlib/constants.rst]: It would be good to clarify here exactly how this differs from `@__FILE__`

**Comment by omus** [comment]
Yes it could be modified. I'll note that `ARGS` elements can also be modified if someone wants to. Alternatively I could make `PROGRAM_FILE` constant without changing anything else. This would mean that if someone tried using `PROGRAM_FILE` in the `--load` file it would raise an UndefVarError.

File: doc/stdlib/constants.rst (hunk 4 above)
Comment (omus) [inline doc/stdlib/constants.rst]: Good point. I'll add some more details.

**Comment by omus** [comment]
@tkelman I updated the documentation to be clearer. I noticed however that macro's don't seem to support permalinks which means the `@__FILE__` link I included in the `PROGRAM_FILE` documentation doesn't work.

File: doc/stdlib/file.rst [hunk 5]
```diff
@@ -320,7 +320,7 @@
 
    .. Docstring generated from Julia source
 
-   ``@__FILE__`` expands to a string with the absolute path and file name of the script being run. Returns ``nothing`` if run from a REPL or an empty string if evaluated by ``julia -e <expr>``\ .
+   ``@__FILE__`` expands to a string with the absolute path and file name of the containing the macro. Returns ``nothing`` if run from a REPL or an empty string if evaluated by ``julia -e <expr>``\ . Alternatively see :data:`PROGRAM_FILE`.
```
Comment (tkelman) [inline doc/stdlib/file.rst]: this is autogenerated so should be changed at the jl docstring source

**Comment by omus** [comment]
I tracked down the permalink issue with `@__FILE__`. Will be fixed when https://github.com/JuliaLang/JuliaDoc/pull/18 is merged. No changes will be required in this PR.

**Comment by omus** [comment]
Rebased on to latest master. I also separated the PROGRAM_FILE tests from the ARGS tests.

**Comment by JeffBezanson** [comment]
LGTM. @StefanKarpinski what do you think?

**Comment by stevengj** [comment]
LGTM.

**Comment by omus** [comment]
Rebased again on to the latest master. Fixed conflict with 7d5def6

**Comment by omus** [comment]
The appveyor 64-bit Windows failure looks like it just got killed at 1 hour.

**Comment by omus** [comment]
Thanks!

File: NEWS.md (hunk 2 above)
Comment (stevengj) [inline NEWS.md]: Hey, I just noticed that this was accidentally added to the Julia 0.4 NEWS, not the Julia 0.5 NEWS...

Comment (stevengj) [inline NEWS.md]: Fixed in 47c29a9401bbd1d7e45e54b80c1b403bf8d3cb8e

**Pull request merged**